### PR TITLE
Clean up the fstarlib mess

### PR DESCRIPTION
**Pull request by ad-l**
fstarlib should only contain F* standard libraries - in particular I am strongly opposed to having F* typechecker modules there as we are linking it with F* applications. I also reduce the package footprint of fstarlib to batteries and zarith, as the other packages are only useful for the compiler itself.

**N.B. This has high priority because miTLS cannot build with the current fstarlib in master due to module name conflicts with compiler-libs.common**

### CODE REVIEW

**Comment by s-zanella** [comment]
Looks good.
Note that the only conflict right now that stops mitls-fstar from compiling is `Parse`.
As a stopgap measure I just renamed it to `TLSParse` in one of my branches.

File: ulib/ml/Makefile [hunk 1]
```diff
@@ -19,17 +19,17 @@ fstarlib.mllib: .mgen *.ml $(wildcard extracted/*.ml)
...
 FSTARLIB_DIR=../../bin/fstarlib/
 
-fstarlib.cma: export OCAMLFIND_IGNORE_DUPS_IN = $(shell ocamlfind query compiler-libs)
+# fstarlib.cma: export OCAMLFIND_IGNORE_DUPS_IN = $(shell ocamlfind query compiler-libs)
 fstarlib.cma:fstarlib.mllib
 	$(OCAMLBUILD) $@
 	cp _build/ulib/ml/$@ $(FSTARLIB_DIR)
 
-fstarlib.cmxa: export OCAMLFIND_IGNORE_DUPS_IN = $(shell ocamlfind query compiler-libs)
+#fstarlib.cmxa: export OCAMLFIND_IGNORE_DUPS_IN = $(shell ocamlfind query compiler-libs)
```
Comment (cpitclaudel) [inline ulib/ml/Makefile]: No need to comment these out. Just remove them.

File: ulib/ml/Makefile [hunk 2]
```diff
@@ -45,7 +45,8 @@ install: fstarlib.cmxa fstarlib.cma
 
 # Common rules
 clean:
-	rm -f fstarlib.mllib
-	rm -f $(ULIB_GEN_DIR)/*
+	rm -f fstarlib.mllib *~
```
Comment (cpitclaudel) [inline ulib/ml/Makefile]: is this intended to remove Emacs temp files?

File: ulib/Makefile [hunk 3]
```diff
@@ -41,23 +41,28 @@ wc:
...
+	cp ../src/extraction/ml/FStar_Extraction_ML_PrintML.ml ml/compiler-extracted/
+	cp ../src/parser/ml/FStar_Parser_LexFStar.mll          ml/compiler-extracted/
+	cp ../src/parser/ml/FStar_Parser_ParseIt.ml            ml/compiler-extracted/
+	cp ../src/parser/ml/FStar_Parser_Util.ml               ml/compiler-extracted/
+	cp ../src/tactics/ml/FStar_Tactics_Native.ml           ml/compiler-extracted/
+	cp ../src/tactics/ml/FStar_Tactics_Load.ml             ml/compiler-extracted/
+	cp ../src/tactics/ml/FStar_Tactics_Builtins.ml         ml/compiler-extracted/
+	$(FSTAR) --lax --codegen OCaml --odir ml/extracted $(FSTAR_DEFAULT_ARGS) $(EXTRACT_FSTARLIB)
+	$(FSTAR) --lax --codegen OCaml --odir ml/compiler-extracted $(FSTAR_DEFAULT_ARGS) $(EXTRACT_COMPILER)
+	cp ml/compiler-extracted/FStar_Pervasives.ml ml/extracted
```
Comment (cpitclaudel) [inline ulib/Makefile]: I think @victor-dumitrescu and I will rewrite this in a few weeks.

File: ulib/Makefile [hunk 4]
```diff
@@ -41,23 +41,28 @@ wc:
 include ml/Makefile.include
 
 REALIZED=$(addsuffix .fst, $(NOEXTRACT_MODULES))
-EXTRACT_ALL=$(filter-out $(REALIZED),$(wildcard FStar.*.fst FStar.*.fsti))
+EXTRACT_FSTARLIB=$(filter-out $(REALIZED),$(filter-out FStar.Crypto.fst FStar.Tcp.fst FStar.Reflection.% FStar.Tactics.%,$(wildcard FStar.*.fst FStar.*.fsti)))
```
Comment (cpitclaudel) [inline ulib/Makefile]: Any reason to nest the `filter-out`s?

File: _tags [hunk 5]
```diff
@@ -9,12 +9,21 @@
 true: \
   annot, bin_annot, thread, -traverse, \
   package(batteries), \
+  package(zarith)
+#  package(dynlink), \
+#  package(pprint), \
+#  package(yojson), \
```
Comment (cpitclaudel) [inline _tags]: No need to comment these out; just remove them

File: _tags [hunk 6]
```diff
@@ -9,12 +9,21 @@
...
   package(compiler-libs), \
   package(compiler-libs.common), \
   package(dynlink), \
   package(pprint), \
-  package(yojson), \
-  package(zarith)
+  package(yojson)
 
 # Turn off warnings for extracted files
 <src/ocaml-output/**/*.ml> or "src/tactics/ml/FStar_Reflection_Syntax.ml" or  "src/tactics/ml/FStar_Tactics_Effect.ml" or "src/tactics/ml/FStar_Reflection_Formula.ml" or <ulib/ml/extracted/*.ml>: \
```
Comment (cpitclaudel) [inline _tags]: Doesn't this rule need to be extended to include `compiler-extracted`?

**Comment by cpitclaudel** [comment]
> in particular I am strongly opposed to having F* typechecker modules there as we are linking it with F* applications

I think everyone agrees :) The way I see it, fstarlib is the thing that will (at some point in the future) end up on OPAM as a dependency of most software written in F*.

**Pull request merged**